### PR TITLE
Implementation of the paper "Jointly Learning to Align & Translate with Transformer"

### PR DESCRIPTION
**Pull request by Zenglinxiao**
Implementation of the paper https://arxiv.org/abs/1909.02074 using OpenNMT-py.
Original implementation disponible at fairseq.
This PR enables Transformer based models to jointly output word alignment alongside its translation.
The quality of output alignment could be further improved by joint training models on translation & alignment providing align labels produced by statistic tools such as fastalign or giza.

### CODE REVIEW

File: onmt/decoders/transformer.py [hunk 1]
```diff
@@ -43,10 +44,34 @@ def __init__(self, d_model, heads, d_ff, dropout, attention_dropout,
...
+            # layer average attention across heads, get ``(B, Q, K)``
+            # Case 1: no full_context, no align heads -> layer avg baseline
+            # Case 2: no full_context, 1 align heads -> guided align
+            # Case 3: full_context, 1 align heads -> full cte guided align
+            attn_align = attns.mean(dim=1)
+        return output, top_attn, attn_align
+
+    def _forward(self, inputs, memory_bank, src_pad_mask, tgt_pad_mask,
+                 layer_cache=None, step=None, future=False):
+        """ A naive forword pass for transformer decoder.
```
Comment (francoishernandez) [inline onmt/decoders/transformer.py]: typo forword --> forward

File: onmt/decoders/transformer.py [hunk 2]
```diff
@@ -43,10 +44,34 @@ def __init__(self, d_model, heads, d_ff, dropout, attention_dropout,
...
+        # TODO: change 1 to T as T could be 1 or tgt_len
```
Comment (francoishernandez) [inline onmt/decoders/transformer.py]: status?

File: onmt/inputters/inputter.py [hunk 3]
```diff
@@ -84,6 +125,7 @@ def get_fields(
             for tgt.
         dynamic_dict (bool): Whether or not to include source map and
             alignment fields.
+        guide_align (bool): Whether or not to include word align.
```
Comment (francoishernandez) [inline onmt/inputters/inputter.py]: maybe use `with_align` as in other parts of the code?

File: onmt/opts.py [hunk 4]
```diff
@@ -199,12 +214,17 @@ def preprocess_opts(parser):
...
+    group.add('--train_align', '-train_align', nargs='+', default=[None],
+              help="Path(s) to the training src-tgt alignment")
     group.add('--train_ids', '-train_ids', nargs='+', default=[None],
               help="ids to name training shards, used for corpus weighting")
+
     group.add('--valid_src', '-valid_src',
               help="Path to the validation source data")
     group.add('--valid_tgt', '-valid_tgt',
               help="Path to the validation target data")
+    group.add('--valid_align', '-valid_align', default=None,
```
Comment (francoishernandez) [inline onmt/opts.py]: Do we have any validation metric on alignments?

File: onmt/opts.py [hunk 5]
```diff
@@ -603,6 +622,8 @@ def translate_opts(parser):
     group.add('--output', '-output', default='pred.txt',
               help="Path to output the predictions (each line will "
                    "be the decoded sequence")
+    group.add('--report_align', '-report_align', action='store_true',
```
Comment (francoishernandez) [inline onmt/opts.py]: How will this work? Will it be dumped in the translation file (`-output`) or do we maybe need to add another file `-output_align`?

File: onmt/translate/translator.py [hunk 6]
```diff
@@ -517,10 +540,90 @@ def _translate_random_sampling(
...
+        full_tgt = torch.cat((bos_tensor, paded_tgt), dim=-1)
+        batch_best_tgt = full_tgt.view(
+            len(predictions), -1, full_tgt.size(-1))  # (batch, n_best, tgt_l)
+        padding_mask = batch_best_tgt.eq(self._tgt_pad_idx)
+        eos_mask = batch_best_tgt.eq(self._tgt_eos_idx)
+        bos_mask = batch_best_tgt.eq(self._tgt_bos_idx)
+        tgt_mask = padding_mask | eos_mask | bos_mask
+        return batch_best_tgt, tgt_mask
+
+    def _align_forword(self, batch, predictions):
```
Comment (francoishernandez) [inline onmt/translate/translator.py]: forword --> forward

File: onmt/translate/translator.py [hunk 7]
```diff
@@ -517,10 +540,90 @@ def _translate_random_sampling(
                     lambda state, dim: state.index_select(dim, select_indices))
 
         results["scores"] = random_sampler.scores
-        results["predictions"] = random_sampler.predictions
+        results["predictions"] = random_sampler.predictions  # (batch, )
         results["attention"] = random_sampler.attention
+        if self.report_align:
+            results["alignment"] = self._align_forword(
```
Comment (francoishernandez) [inline onmt/translate/translator.py]: forword --> forward

File: onmt/translate/translator.py [hunk 8]
```diff
@@ -719,8 +822,13 @@ def _translate_batch(
                 lambda state, dim: state.index_select(dim, select_indices))
 
         results["scores"] = beam.scores
-        results["predictions"] = beam.predictions
+        results["predictions"] = beam.predictions  # (batch, n_best)
         results["attention"] = beam.attention
+        if self.report_align:
+            results["alignment"] = self._align_forword(
```
Comment (francoishernandez) [inline onmt/translate/translator.py]: forword --> forward

File: onmt/utils/loss.py [hunk 9]
```diff
@@ -248,10 +250,30 @@ def _make_shard_state(self, batch, output, range_, attns=None):
                 "std_attn": attns.get("std"),
                 "coverage_attn": coverage
             })
+        if self.lambda_align != 0.0:
+            # attn_align should be in (batch_size, pad_tgt_size, pad_src_size)
+            attn_align = attns.get("align", None)
+            # align_idx should be a Tensor in size([N, 3]) of [B, tgt+1, src]
```
Comment (francoishernandez) [inline onmt/utils/loss.py]: ?

File: onmt/utils/loss.py [hunk 10]
```diff
@@ -248,10 +250,30 @@ def _make_shard_state(self, batch, output, range_, attns=None):
...
+            assert attn_align is not None, "lambda_align != 0.0 requires " \
+                "alignement attention head"
+            assert align_idx is not None, "lambda_align != 0.0 requires " \
+                "provide guided alignement"
+            pad_tgt_size, batch_size, _ = batch.tgt.size()
+            pad_src_size = batch.src[0].size(0)
+            align_matrix_size = [batch_size, pad_tgt_size, pad_src_size]
+            ref_align = onmt.utils.make_batch_align_matrix(
+                align_idx, align_matrix_size, normalize=True)
+            # NOTE: tgt-src alignement that in range_, not take count in bos
```
Comment (francoishernandez) [inline onmt/utils/loss.py]: ?

File: onmt/utils/misc.py [hunk 11]
```diff
@@ -3,10 +3,21 @@
 import torch
 import random
 import inspect
-from itertools import islice
+from itertools import islice, cycle
 
 
-def split_corpus(path, shard_size):
+def split_corpus(path, shard_size, default=None):
```
Comment (francoishernandez) [inline onmt/utils/misc.py]: not sure to understand the point here

File: onmt/utils/parse.py [hunk 12]
```diff
@@ -63,6 +68,16 @@ def validate_model_opts(cls, model_opt):
             if model_opt.model_type != "text":
                 raise AssertionError(
                     "--share_embeddings requires --model_type text.")
+        if model_opt.lambda_align > 0.0:
+            assert model_opt.decoder_type == 'transformer', \
+                "Only transformer is supported to joint learn alignment."
+            assert model_opt.alignment_layer < model_opt.layers, \
+                "N° alignment_layer should be smaller than number of layers."
+            print("[OPTS] Joint learn alignment at layer'{}'"
```
Comment (francoishernandez) [inline onmt/utils/parse.py]: shouldn't we use `logger`?

**Comment by francoishernandez** [comment]
We need to fix the audio codepath to handle the new options (see Travis unit test logs).
Also, it might be good to have an entry in the FAQ about this: what's the point, how to use (preprocess, train, translate sample commands and data format example).

File: onmt/utils/loss.py (hunk 9 above)
Comment (Zenglinxiao) [reply onmt/utils/loss.py]: batch.align contains alignments in the format of alignment idx rather than alignment matrix. batch.align may contain N pair of src-tgt alignment, each represented as (sentence_number_in_batch, aligned_tgt_id+1, aligned_src_id) . aligned_tgt_id+1 behaves as adding BOS in each sentence, aims to making it coherent with batch.tgt after calling make_batch_align_matrix for batch.align.

File: onmt/utils/loss.py (hunk 10 above)
Comment (Zenglinxiao) [reply onmt/utils/loss.py]: By calling make_batch_align_matrix, the sparse align_idx will be converted to a full alignment matrix. To make it coherent with shard_state['target'], we return a submatrix in the same range of shard_state['target'],.

File: onmt/opts.py (hunk 5 above)
Comment (Zenglinxiao) [reply onmt/opts.py]: We just paste the alignment output behind translation, separated by ` ||| `, ex: 'This is a translation ||| 1-1 2-2 3-3'

File: onmt/opts.py (hunk 4 above)
Comment (Zenglinxiao) [reply onmt/opts.py]: We do the same as for train_align.

File: onmt/decoders/transformer.py (hunk 2 above)
Comment (Zenglinxiao) [reply onmt/decoders/transformer.py]: This will be done in the next PR.

File: onmt/bin/preprocess.py [hunk 13]
```diff
@@ -125,19 +126,21 @@ def maybe_load_vocab(corpus_type, counters, opt):
     return src_vocab, tgt_vocab, existing_fields
 
 
-def build_save_dataset(corpus_type, fields, src_reader, tgt_reader, opt):
+def build_save_dataset(corpus_type, fields, r_src, r_tgt, r_align, opt):
```
Comment (francoishernandez) [inline onmt/bin/preprocess.py]: Why the change from X_reader to r_X? Especially when you're still using X_reader in other parts.

File: onmt/bin/server.py [hunk 14]
```diff
@@ -75,14 +75,17 @@ def translate():
         inputs = request.get_json(force=True)
         out = {}
         try:
-            trans, scores, n_best, times = translation_server.run(inputs)
+            trans, scores, n_best, _, aligns = translation_server.run(inputs)
```
Comment (francoishernandez) [inline onmt/bin/server.py]: Not sure why we returned the times in the first place, since we're logging translation time inside the function already. @pltrdy was there a reason or can we simplify and remove this from the return?

File: onmt/decoders/transformer.py [hunk 15]
```diff
@@ -43,10 +44,34 @@ def __init__(self, d_model, heads, d_ff, dropout, attention_dropout,
         self.layer_norm_1 = nn.LayerNorm(d_model, eps=1e-6)
         self.layer_norm_2 = nn.LayerNorm(d_model, eps=1e-6)
         self.drop = nn.Dropout(dropout)
+        self.full_context_alignment = full_context_alignment
+        self.alignment_heads = alignment_heads
 
-    def forward(self, inputs, memory_bank, src_pad_mask, tgt_pad_mask,
-                layer_cache=None, step=None):
+    def forward(self, *args, **kwargs):
+        """ Extend _forward for multiply decoder pass.
```
Comment (francoishernandez) [inline onmt/decoders/transformer.py]: multiply --> multiple?
Maybe add one sentence to explain why we need it?

File: onmt/decoders/transformer.py [hunk 16]
```diff
@@ -57,45 +82,49 @@ def forward(self, inputs, memory_bank, src_pad_mask, tgt_pad_mask,
...
+                                      layer_cache=layer_cache, step=step)
 
         query = self.drop(query) + inputs
 
         query_norm = self.layer_norm_2(query)
-        mid, attn = self.context_attn(memory_bank, memory_bank, query_norm,
-                                      mask=src_pad_mask,
-                                      layer_cache=layer_cache,
-                                      attn_type="context")
+        mid, attns = self.context_attn(memory_bank, memory_bank, query_norm,
```
Comment (francoishernandez) [inline onmt/decoders/transformer.py]: mid?

File: onmt/decoders/transformer.py [hunk 17]
```diff
@@ -148,7 +178,9 @@ def __init__(self, num_layers, d_model, heads, d_ff,
             [TransformerDecoderLayer(d_model, heads, d_ff, dropout,
              attention_dropout, self_attn_type=self_attn_type,
              max_relative_positions=max_relative_positions,
-             aan_useffn=aan_useffn)
+             aan_useffn=aan_useffn,
+             full_context_alignment=full_context,
```
Comment (francoishernandez) [inline onmt/decoders/transformer.py]: Why two different names?

File: onmt/decoders/transformer.py [hunk 18]
```diff
@@ -232,6 +275,9 @@ def forward(self, tgt, memory_bank, step=None, **kwargs):
         attns = {"std": attn}
         if self._copy:
             attns["copy"] = attn
+        if with_align:
+            attns["align"] = attn_aligns[self.alignment_layer]  # `(B, Q, K)`
+            # attns["align"] = torch.stack(attn_aligns, 0).mean(0)  # All avg
```
Comment (francoishernandez) [inline onmt/decoders/transformer.py]: Can't we make this an option to allow easier experimentation? Maybe pass a specific value to one of the options you introduced?

File: onmt/translate/translation_server.py [hunk 19]
```diff
@@ -440,19 +441,23 @@ def run(self, inputs):
...
         results = flatten_list(predictions)
         scores = [score_tensor.item()
                   for score_tensor in flatten_list(scores)]
 
-        results = [self.maybe_detokenize(item)
-                   for item in results]
+        results = [self.maybe_detokenize_with_align(result, src)
+                   for result, src in zip(results, tiled_texts)]
+
+        aligns = [align for _, align in results]
```
Comment (francoishernandez) [inline onmt/translate/translation_server.py]: Do we always have some alignments in the predictions?

File: onmt/translate/translator.py [hunk 20]
```diff
@@ -382,17 +400,23 @@ def translate(
...
-                    for word, row in zip(preds, attns):
-                        max_index = row.index(max(row))
-                        row_format = row_format.replace(
-                            "{:>10.7f} ", "{:*>10.7f} ", max_index + 1)
-                        row_format = row_format.replace(
-                            "{:*>10.7f} ", "{:>10.7f} ", max_index)
-                        output += row_format.format(word, *row) + '\n'
-                        row_format = "{:>10.10} " + "{:>10.7f} " * len(srcs)
+                    output = report_matrix(srcs, preds, attns)
+                    if self.logger:
```
Comment (francoishernandez) [inline onmt/translate/translator.py]: Are there some cases where we don't have a logger?

File: onmt/translate/translator.py [hunk 21]
```diff
@@ -517,10 +541,90 @@ def _translate_random_sampling(
                     lambda state, dim: state.index_select(dim, select_indices))
 
         results["scores"] = random_sampler.scores
-        results["predictions"] = random_sampler.predictions
+        results["predictions"] = random_sampler.predictions  # (batch, )
         results["attention"] = random_sampler.attention
+        if self.report_align:
+            results["alignment"] = self._align_forward(
```
Comment (francoishernandez) [inline onmt/translate/translator.py]: Aren't we supposed to use the already performed forward to get alignments? I reckon this second pass would hurt inference time a lot, wouldn't it?

File: onmt/bin/preprocess.py (hunk 13 above)
Comment (Zenglinxiao) [reply onmt/bin/preprocess.py]: It's just for shortening the line to not exceed 79 required by flake.

File: onmt/decoders/transformer.py (hunk 16 above)
Comment (Zenglinxiao) [reply onmt/decoders/transformer.py]: I guess this naming is just because context attention block is at the middle of the decoder (self-attention -> context-attention -> feed forward) block.

File: onmt/translate/translation_server.py (hunk 19 above)
Comment (Zenglinxiao) [reply onmt/translate/translation_server.py]: For each prediction, we return a align or None.

File: onmt/translate/translator.py (hunk 20 above)
Comment (Zenglinxiao) [reply onmt/translate/translator.py]: Exactly! I just verified that the logger has not correctly initialized. As in both call of `build_translator` in `translate.py` and `translation_server.py`, the argument logger has not been passed in. Therefore, we actually only call os.write. The same for the method self._log.
Maybe we could create another separate PR to fix them?

File: onmt/translate/translator.py (hunk 21 above)
Comment (Zenglinxiao) [reply onmt/translate/translator.py]: I think this is necessary for full_context_alignment, as in decoder time we generate translation regressively, we can only perform future aware decode after having the prediction.
And this will hurt inference a little bit rather than a lot, as most of the time is used during beam search, this is just a forward pass to get the inner hidden states(attention).

File: onmt/decoders/transformer.py (hunk 17 above)
Comment (Zenglinxiao) [reply onmt/decoders/transformer.py]: Aims to shorten the sentence of line 169 and meet the requirement of flake (79).

File: onmt/decoders/transformer.py (hunk 15 above)
Comment (Zenglinxiao) [reply onmt/decoders/transformer.py]: How about this:
""" Extend _forward for (possibly) multiple decoder pass:
        1. Always a default (future masked) decoder forward pass,
        2. Possibly a second future aware decoder pass for joint learn
            full context alignement.
"""

File: onmt/translate/translator.py [hunk 22]
```diff
@@ -428,6 +452,79 @@ def translate(
                       codecs.open(self.dump_beam, 'w', 'utf-8'))
         return all_scores, all_predictions
 
+    def _batch_predictions(self, predictions, bos, pad):
```
Comment (francoishernandez) [inline onmt/translate/translator.py]: This name should explicitly refer to the alignment task I think. Also, it's not very clear what this function is for.

Comment (Zenglinxiao) [reply onmt/translate/translator.py]: This function aims to rebuild the tgt tensor from predictions.
The predictions is n_batch List of n_best list containing generated predictions ended with eos index.
To achieve that, we need to do padding and add bos index.
It's a function that kind similar to pad() in torchtext.Field.

Comment (Zenglinxiao) [reply onmt/translate/translator.py]: maybe naming this function as _align_pad_prediction ?

Comment (francoishernandez) [reply onmt/translate/translator.py]: Seems good to me.
